OData.Community.AlternateKeys.V1.md

```diff
@@ -0,0 +1,314 @@
+﻿#Alternate Keys for OData Services
+##A mechanism for OData services to express additional unique key definitions that may be used to address entities.
+
+##1	Introduction
+###1.1	Summary
+This document details a proposal for an OData v4 vocabulary to describe a services ability to have certain entities be addressed by their standard key properties as well as additional sets of properties that make up alternate keys.
+
+###1.2	Background
+OData v4 currently includes the concept of a key for an entity type that defines the set of columns that make up the unique identifier for an entity.  This key is used in the URL that addresses a specific entity instance.  For example:
+
+```http
+https://host/service/customers(123)
+https://host/service/customers(CustomerId=123)
+```
+
+A key can be composed of multiple properties.  For example 
+
+```http
+https://host/service/customers(Branch=ABC,CustomerId=123)
+```
+
+For some data models there can be multiple ways to uniquely identify an entity.  A customer might be identified by an application’s CustomerId property as a primary key as well as by an enterprise customer master id and an industry standard DUNS number.  Different data integration scenarios can be achieved more efficiently if entities can be addressed using these keys directly.  For example
+
+```http
+https://host/service/customers(customermasterid=987)
+https://host/service/customers(DUNSNumber=665544332211)
+https://host/service/customers(Branch=ABC,CustomerId=123)
+```
+
+This concept is missing from current OData specifications.
+
+###1.3	Further Reading
+OData v4 specifications:  http://www.odata.org/documentation/
+ 
+##2	Usage Scenarios
+###2.1	Client requests service metadata
+When metadata-aware clients first retrieve metadata from a service to discover its data model, entity type metadata will be annotated to describe the availability of alternate keys that can be used to address that entity type.
+
+####2.1.1	Alternate Keys Vocabulary
+When a service supports alternate keys, it will annotate entity type metadata to describe the structure of the available alternate keys for each entity type.
+
+The AlternateKeys annotation term will contain a collection of key definitions, each of which contain a collection of PropertyRef’s.  It will be declared as follows:
+
+```xml
+<Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="OData.Community.AlternateKeys.V1" Alias="Keys">
+
+  <Term AppliesTo="EntityType EntitySet NavigationProperty" Type="Collection(Keys.AlternateKey)" Name="AlternateKeys">
+    <Annotation Term="Core.Description" String="Communicates available alternate keys"/>
+  </Term>
+
+  <ComplexType Name="AlternateKey">
+    <Property Type="Collection(Keys.PropertyRef)" Name="Key">
+      <Annotation Term="Core.Description" String="The set of properties that make up this key"/>
+    </Property>
+  </ComplexType>
+
+  <ComplexType Name="PropertyRef">
+    <Property Type="Edm.PropertyPath" Name="Name">
```

Comment:
Just a side note that: EdmLib for now might not support PropertyPath yet.  Related issue: https://github.com/OData/odata.net/issues/37